File: Algorithm/Clone Graph.java
/**
 * Definition for undirected graph.
 * class UndirectedGraphNode {
 *     int label;
 *     List<UndirectedGraphNode> neighbors;
 *     UndirectedGraphNode(int x) { label = x; neighbors = new ArrayList<UndirectedGraphNode>(); }
 * };
 */
public class Solution {
    public UndirectedGraphNode cloneGraph(UndirectedGraphNode node) {
        if (node == null) {
        	return node;
        }
        Map<UndirectedGraphNode, UndirectedGraphNode> map = new HashMap<UndirectedGraphNode, UndirectedGraphNode>();
        Queue<UndirectedGraphNode> queue = new LinkedList<UndirectedGraphNode>();
        map.put(node, new UndirectedGraphNode(node.label));
        queue.offer(node);
        while (!queue.isEmpty()) {
        	UndirectedGraphNode oldNode = queue.poll();
        	UndirectedGraphNode newNode = map.get(oldNode);
        	for (UndirectedGraphNode neighbor : oldNode.neighbors) {
        		if (!map.containsKey(neighbor)) {
        			queue.offer(neighbor);
        			map.put(neighbor, new UndirectedGraphNode(neighbor.label));
        		}
        		newNode.neighbors.add(map.get(neighbor));
        	}
        }
        return map.get(node);
    }
}

File: Algorithm/JavaScript/Clone Graph.js
/**
 * // Definition for a Node.
 * function Node(val, neighbors) {
 *    this.val = val === undefined ? 0 : val;
 *    this.neighbors = neighbors === undefined ? [] : neighbors;
 * };
 */

/**
 * @param {Node} node
 * @return {Node}
 */
var cloneGraph = function(node) {
    if (!node) {
        return node;
    }
    const map = {};

    const helper = (node) => {
        if (!node) {
            return null;
        }
        if (!map[node.val]) {
            map[node.val] = new Node(node.val);
            map[node.val].neighbors = node.neighbors.map(helper);
        }
        return map[node.val];
    }

    return helper(node);
};